dask/layers.py

```diff
@@ -132,13 +132,72 @@ def _dict(self):
         return self._cached_dict
 
     def __getitem__(self, key):
-        return self._dict[key]
+        """Fetch / materialize a single item"""
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict[key]
+
+        # Raise a TypeError if the key isn't hashable
+        set(key)
+
+        # If the first value isn't a string, it's not valid
+        if not hasattr(key, "__len__") or (len(key) < 1) or not isinstance(key[0], str):
+            raise KeyError
+
+        getitem_name = f"getitem-{self.token}"
+        overlap_name = f"overlap-{self.token}"
+
+        if key[0] == overlap_name:
+            from dask.array.core import concatenate3
+
+            return (
+                concatenate3,
+                (
+                    concrete,
+                    _expand_keys_around_center(
+                        key, self.numblocks, getitem_name, self.axes
+                    ),
+                ),
+            )
+
+        if key[0] == getitem_name:
+            rounded = (self.name,) + tuple(round(k) for k in key[1:])
+            if rounded[1:] == tuple(key[1:]):
+                return rounded
+            return fractional_slice((self.name,) + key[1:], self.axes)
+        raise KeyError
 
     def __iter__(self):
         return iter(self._dict)
 
     def __len__(self):
-        return len(self._dict)
+        """This could be calculated directly to remove the numpy dependency
+        and speed up the calculation slightly, but have only expanded it
+        to two block dimensions so far:
+            size = lambda a,b: (
+                (a*b) +                   # Overlap blocks; np.prod(blocks)
+                ((a-2)*(b-2))*(3*3)     + # No masked values, in middle
+                ((a-2)+(b-2))*(3*2) * 2 + # One masked value, both sides
+                2*(2*2) * 2               # One masked for each, both sides
+            )
+        """
+        if getattr(self, "_cached_len", None) is None:
```

Comment:
I'm not sure how carefully I will be able to review/check this :)

We will definitely want to make sure we have tests that confirm we are getting correct results for a variety of different map-overlap examples.

Reply:
It's basically performing a vectorized version of `_expand_keys_around_center`, but only looking at whether the keys exist via the mask. Agreed on the testing - can write a test that checks that `len(ArrayOverlapLayer) ==  len(ArrayOverlapLayer._dict)` for a variety of examples.